Sofa/framework/Core/src/sofa/core/behavior/ForceField.inl

```diff
@@ -55,17 +55,46 @@ void ForceField<DataTypes>::addDForce(const MechanicalParams* mparams, MultiVecD
 {
     if (mparams && this->mstate)
     {
-
-#ifndef NDEBUG
-            mparams->setKFactorUsed(false);
-#endif
-
-        addDForce(mparams, *dfId[this->mstate.get()].write(), *mparams->readDx(this->mstate.get()));
-
-#ifndef NDEBUG
-        if (!mparams->getKFactorUsed())
-            msg_warning()  << getClassName() << " (in ForceField<DataTypes>::addDForce): please use mparams->kFactor() in addDForce";
-#endif
+        const auto dx = sofa::helper::getReadAccessor(*mparams->readDx(this->mstate.get()));
+        auto df = sofa::helper::getWriteOnlyAccessor(*dfId[this->mstate.get()].write());
+        const Real kFactor = static_cast<Real>(
+            sofa::core::mechanicalparams::kFactorIncludingRayleighDamping(mparams, this->rayleighStiffness.getValue()));
+
+        /// Instead of accumulating the matrix contributions in a matrix data structure, this class
+        /// directly multiply the contribution by the dx vector
+        class MatrixFreeAccumulator : public StiffnessMatrixAccumulator
+        {
+            const VecDeriv& m_dx;
+            VecDeriv& m_df;
+            const Real m_kFactor;
+
+        public:
+            MatrixFreeAccumulator(const VecDeriv& dx, VecDeriv& df, const Real kFactor) : m_dx(dx), m_df(df), m_kFactor(kFactor) {}
+
+            void add(sofa::SignedIndex row, sofa::SignedIndex col, float value) override
```

Comment:
My overhead detection sub-system is emitting a beep for these lines.

Reply:
Where did you get this feature? 😄

Reply:
Is it because of the virtuality?

Reply:
I would says that this is triggered by: individualized data access (row,col) + virtuality call overhead + virtuality preventing the compiler to inline the add function. Not being able to inline will prevent the compiler to detect the relationship between successives calls to dfId, dxId and thus will imply, each time, to recompute them. 

As always when overhead detection sub-system is emitting a beep, Godbolt... must be used to confirm there is really an issue :)